src/main/java/core/basesyntax/strategy/impl/PurchaseOperation.java

```
@@ -0,0 +1,21 @@
+package core.basesyntax.strategy.impl;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.strategy.OperationHandler;
+
+public class PurchaseOperation implements OperationHandler {
+    @Override
+    public void handle(FruitTransaction transaction) {
+        if (transaction == null) {
+            throw new RuntimeException("Transaction cannot be null.");
+        }
+        int currentQuantity = Storage.getFruitQuantity(transaction.getFruit());
+        int newQuantity = currentQuantity - transaction.getQuantity();
```

You use `transaction.getQuantity()` at line 14 without validating it. Even though DataConverter enforces non-negative quantities, handlers are public and should be defensive. Add an explicit check such as `if (transaction.getQuantity() < 0) { throw new RuntimeException("Quantity cannot be negative: " + transaction.getQuantity()); }` before computing newQuantity so invalid transactions fail fast with an informative message. This aligns with: "Remember that your code will need to be tested, so try to anticipate and handle all invalid input data in advance."